### PR TITLE
Give autocomplete providers the room they're in

### PR DESCRIPTION
**Pull request by dbkr**
Removes the gut-wrenching that RoomView does to jam the user list
into the user autocomplete provider.

### CODE REVIEW

File: src/autocomplete/Autocompleter.js [hunk 1]
```diff
@@ -43,43 +45,59 @@ const PROVIDERS = [
...
 
-            /* the currently matched "command" the completer tried to complete
-             * we pass this through so that Autocomplete can figure out when to
-             * re-show itself once hidden.
-             */
-            command: PROVIDERS[i].getCurrentCommand(query, selection, force),
-        };
-    });
+    async getCompletions(query: string, selection: SelectionRange, force: boolean = false): Array<Completion> {
+        /* Note: That this waits for all providers to return is *intentional*
```
Comment (lukebarnard1) [inline src/autocomplete/Autocompleter.js]: Needs a bit of Englishing

File: src/components/views/rooms/Autocomplete.js [hunk 2]
```diff
@@ -41,6 +60,11 @@ export default class Autocomplete extends React.Component {
     }
 
     componentWillReceiveProps(newProps, state) {
+        if (this.props.room.roomId !== newProps.room.roomId) {
+            this.autocompleter.destroy();
+            this.autocompleter = new Autocompleter();
```
Comment (lukebarnard1) [inline src/components/views/rooms/Autocomplete.js]: I'm guessing this should have `newProps.room` going into it?

File: src/components/views/rooms/Autocomplete.js [hunk 3]
```diff
@@ -267,8 +295,11 @@ export default class Autocomplete extends React.Component {
...
     // the query string for which to show autocomplete suggestions
-    query: React.PropTypes.string.isRequired,
+    query: PropTypes.string.isRequired,
 
     // method invoked with range and text content when completion is confirmed
-    onConfirm: React.PropTypes.func.isRequired,
+    onConfirm: PropTypes.func.isRequired,
+
+    // The room in which we're autocompleting
+    room: PropTypes.object,
```
Comment (lukebarnard1) [inline src/components/views/rooms/Autocomplete.js]: `instanceOf(MatrixRoom)` maybe?

File: src/autocomplete/AutocompleteProvider.js [hunk 4]
```diff
@@ -28,6 +28,9 @@ export default class AutocompleteProvider {
         }
     }
 
+    destroy() {
```
[the remark this answers is not in the document]
Comment (dbkr) [reply src/autocomplete/AutocompleteProvider.js]: It's a stub so the subclasses don't have to implement it if they don't want to

[the remark this answers is not in the document]
Comment (lukebarnard1) [reply src/autocomplete/AutocompleteProvider.js]: Maybe just a "this is a stub" comment, then?

File: src/autocomplete/UserProvider.js [hunk 5]
```diff
@@ -30,20 +30,54 @@ import type {Room, RoomMember} from 'matrix-js-sdk';
...
+        MatrixClientPeg.get().on("Room.timeline", this._onRoomTimelineBound);
+        MatrixClientPeg.get().on("RoomState.members", this._onRoomStateMemberBound);
+    }
+
+    destroy() {
+        MatrixClientPeg.get().removeListener("Room.timeline", this._onRoomTimelineBound);
+        MatrixClientPeg.get().removeListener("RoomState.members", this._onRoomStateMemberBound);
+    }
+
+    _onRoomTimeline(ev, room, toStartOfTimeline, removed, data) {
```
[the remark this answers is not in the document]
Comment (dbkr) [reply src/autocomplete/UserProvider.js]: mm, it wasn't but yes, I think you're right

File: src/components/views/rooms/Autocomplete.js (hunk 2 above)
Comment (dbkr) [reply src/components/views/rooms/Autocomplete.js]: ooh, good spot

File: src/autocomplete/AutocompleteProvider.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (dbkr) [reply src/autocomplete/AutocompleteProvider.js]: done

**Pull request merged**